File: 006.ZigZagConversion/zigzag.cc
#include <string>
using namespace std;
class Solution {
public:
    string convert(string s, int numRows) {
        string zigzag[numRows];

        int row = 0;
        bool down = true;
        for (char c : s) {
            zigzag[row] += c;

            // 判断方向
            if (down && row + 1 == numRows) {
                down = false;
            }
            if (!down && row - 1 < 0) {
                down = true;
            }

            // 判断是否增减行号
            if (down && row + 1 < numRows) {
                row++;
            }
            if (!down && row - 1 >= 0) {
                row--;
            }
        }

        string ans;
        for (string sb : zigzag) {
            ans += sb;
        }

        return ans;
    }
};

int main(int argc, char const *argv[]) {
    Solution sl = Solution();
    string s = "AB";
    cout << sl.convert(s, 1);
    return 0;
}

File: 006.ZigZagConversion/zigzag.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        zigzagArr = ['' for row in range(numRows)]
        row = 0
        down = True
        for c in s:
            zigzagArr[row] += c

            # 判断方向
            if down and row + 1 == numRows:
                down = False
            if not down and row - 1 < 0:
                down = True

            # 判断是否增减行号
            if down and row + 1 < numRows:
                row += 1
            if not down and row - 1 >= 0:
                row -= 1

        ans = ''
        for row in zigzagArr:
            ans += row

        return ans


if __name__ == '__main__':
    s = 'AB'
    numRows = 1
    print Solution().convert(s, numRows)
